File: Hashmap/Q3.Group Anagrams/AvwalKaur_Q3_GroupAnagrams.java
import java.util.*;

public class AvwalKaur_Q3_GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] arr) {
        List<List<String>> ans=new ArrayList<>();
        if(arr.length==0) return ans; // return an empty arraylist
        
        // for every word in the array, if we sort that word and corresponding to that we store the 
        // actual words, then we will be able to store the anagrams at one plsce
        Map<String, List<String>> mpp=new HashMap<>();
        for(String s:arr) {
            char[] word=s.toCharArray();
            Arrays.sort(word);
            String sortedWord=String.valueOf(word);

            if(!mpp.containsKey(sortedWord)) mpp.put(sortedWord, new ArrayList<>());
            // if the word is already present push the new string in arraylist
            mpp.get(sortedWord).add(s);
        }
        return new ArrayList<>(mpp.values());
    }
}

File: Hashmap/Q3.Group Anagrams/AvwalKaur_Q3_GroupAnagrams.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<algorithm>
using namespace std;

vector<vector<string>> groupAnagrams(vector<string> &arr) {
    vector<vector<string>> ans;
    if(arr.size()==0) return ans;

    unordered_map<string, vector<string>> mpp;
    for(auto x:arr) {
        string word=x;
        sort(word.begin(), word.end());
        mpp[word].push_back(x);
    }

    for(auto x:mpp) ans.push_back(x.second);
    return ans;
}

int main() {
    vector<string> arr={"eat", "tan", "nat", "ate", "tea", "bat"};
    vector<vector<string>> ans=groupAnagrams(arr);
    for(auto row:ans) {
        for(auto col:row) {
            cout<<col<<" ";
        }
        cout<<endl;
    }
    return 0;
}